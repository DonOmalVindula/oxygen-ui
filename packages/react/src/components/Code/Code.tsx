/**
 * Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import type {OverridableComponent} from '@mui/material/OverridableComponent';
import clsx from 'clsx';
import {forwardRef} from 'react';
import type {ElementType, Ref, ReactElement} from 'react';
import type {TypographyTypeMap} from '../Typography';
import Typography, {TypographyProps} from '../Typography/Typography';
import './code.scss';

export type CodeProps<C extends ElementType = ElementType> = TypographyProps<C> & {
  /**
   * Shows the code block with a filled background.
   * @default true
   */
  filled?: boolean;
  /**
   * Renders the code block with an outline.
   * @default false
   */
  outlined?: boolean;
};

/**
 * The Code can represent an inline or block code without syntax highlight.
 *
 * Demos:
 *
 * - [Code (Oxygen UI)](https://wso2.github.io/oxygen-ui/react/?path=/docs/data-display-code)
 *
 * API:
 *
 * - inherits [Typography API](https://mui.com/material-ui/api/typography/)
 *
 * @remarks
 * - ✨ This is a custom component that is not available in the Material-UI library.
 * - ✔️ Props of the [Typography](https://mui.com/material-ui/api/typography/) component are also available.
 * - ✅ `component` prop is supported.
 * - ✅ The `ref` is forwarded to the root element.
 *
 * @template C - The type of the component.
 * @param props - The props for the Code component.
 * @param ref - The ref to be forwarded to the Typography component.
 * @returns The rendered Code component.
 */
const Code: OverridableComponent<TypographyTypeMap<CodeProps>> = forwardRef(
  <C extends ElementType = ElementType>(
    {className, children, filled = true, outlined = false, ...rest}: CodeProps<C>,
    ref: Ref<HTMLDivElement>,
  ): ReactElement => {
    const classes: string = clsx('oxygen-code', {filled, outlined}, className);

    return (
      <Typography ref={ref} component="code" className={classes} {...rest}>
        {children}
      </Typography>
    );
  },
) as OverridableComponent<TypographyTypeMap<CodeProps>>;

export default Code;
