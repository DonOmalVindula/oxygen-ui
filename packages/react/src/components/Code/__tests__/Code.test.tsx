/**
 * Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {render} from '@unit-testing';
import Code from '../Code';

describe('Code', () => {
  it('should render successfully', () => {
    const {baseElement} = render(<Code>Code Block</Code>);
    expect(baseElement).toBeTruthy();
  });

  it('should match the snapshot', () => {
    const {baseElement} = render(<Code>Code Block</Code>);
    expect(baseElement).toMatchSnapshot();
  });

  it('should have the correct class name', () => {
    const {container} = render(<Code>Code Block</Code>);
    expect(container.firstChild).toHaveClass('oxygen-code');
  });

  it('should have the correct text', () => {
    const {container} = render(<Code>Code Block</Code>);
    expect(container.firstChild).toHaveTextContent('Code Block');
  });

  it('applies the "filled" style when the "filled" prop is true', () => {
    const {container} = render(<Code filled>Code Block</Code>);
    expect(container.firstChild).toHaveClass('filled');
  });

  it('applies the "outlined" style when the "outlined" prop is true', () => {
    const {container} = render(<Code outlined>Code Block</Code>);
    expect(container.firstChild).toHaveClass('outlined');
  });
});
