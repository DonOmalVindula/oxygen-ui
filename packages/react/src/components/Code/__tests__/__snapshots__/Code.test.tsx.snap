// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code should match the snapshot 1`] = `
<body>
  <div>
    <code
      class="MuiTypography-root MuiTypography-body1 oxygen-typography OxygenTypography-root oxygen-code filled OxygenCode-root OxygenCode-filled css-1qp7oek-MuiTypography-root"
    >
      Code Block
    </code>
  </div>
</body>
`;
